Boost: Track Critical CSS generation Statistics

karthikax opened the pull request:
Fixes #20974

#### Description
Changes in this pull request allows us to track the following:

- The size of Critical CSS blocks which are generated,
- The length of time it takes to generate all the Critical CSS for a site,
- How often each type of error comes up,
- How often the process succeeds.

#### Changes proposed in this Pull Request:
- Enqueue tracks javascript library
- Add callback for tracking events triggered via ajax calls.
- Record the starting time of _Generate Critical CSS_ process
- Send the event to Tracks after completion of each Critical CSS block (for both success and failure).

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
NA

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
Yes, the following:
- Time taken for _Generate Critical CSS_ process
- Status of _Generate Critical CSS_ process (`success` or `error`)
- Error type and Error message in case of Generate Critical CSS fails.
- Percentage of the process completed in case the _Generate Critical CSS_ fails before completion.

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to a site with Jetpack Boost Plugin activated.
* Navigate to wp-admin/admin.php?page=jetpack-boost.
* Enable `Optimize CSS Loading` module.
* Once the process completes, navigate to the Tracks dashboard and verify the recorded event as explained here: PCYsg-4D1-p2
* Note: Usually events take about 5 minutes to reflect on the dashboard.
<img width="911" alt="tracks" src="https://user-images.githubusercontent.com/8727183/133391841-ca1e3d3c-2c26-499d-a3db-366360cc2c59.png">

karthikax committed:
Add track script enqueue

karthikax committed:
Add changelog

karthikax committed:
Refactor Tracks to use Analytics class

karthikax committed:
Add custom events for Critical CSS

Record events for both individual and total success and error states of Critical CSS generation via ajax calls.

karthikax committed:
Merge branch 'master' into add/track-critical-css

karthikax committed:
Refactor jpTracksAJAX type definition

Declare jpTracksAJAX as a constant as opposed to extending the Window interface to maintain code consistency.

karthikax committed:
Add basic tests for Analytics

karthikax committed:
Merge branch 'master' into add/track-critical-css

davidlonjon commented:
Hi @karthikax,

This is a great base you have built here and it will become super handy to get the stats. 

I have been testing it locally and I had a few thoughts. They might be off the original objectives so we can discard them if deemed not effective

If we don't take into consideration the size of the CSS generated and the time it took to generate it just yet (please bear with me on these two) I was wondering if it would be better to send to tracks only the final Critical CSS status so we have only one object for each site for each generation. For example, at the moment the final Critical CSS status can look like the following:

If a complete successful generation:

```
{
  "status": "success",
  "success_count": 3,
  "core_providers": [
    "core_front_page",
    "core_posts_page",
    "singular_page",
    "singular_post",
    "singular_product"
  ],
  "core_providers_status": "success",
  "providers_errors": [],
  "provider_key_labels": [],
  "created": 1631793321.300462,
  "viewports": [
    {
      "type": "phone",
      "width": 414,
      "height": 896
    },
    {
      "type": "tablet",
      "width": 1200,
      "height": 800
    },
    {
      "type": "desktop",
      "width": 1920,
      "height": 1080
    }
  ],
  "generation_nonce": "5c2be69a79",
  "callback_passthrough": {
    "_nonce": "9bf70c6de2"
  }
}
```

If a successful generation but with some errors:

```
{
  "status": "success",
  "success_count": 2,
  "core_providers": [
    "core_front_page",
    "core_posts_page",
    "singular_page",
    "singular_post",
    "singular_product"
  ],
  "core_providers_status": "error",
  "providers_errors": {
    "singular_post": {
      "http://jetpack.jp.ngrok.io/id-suscipit-aspernatur-nihil/": {
        "message": "Failed to fetch cross-domain content at http://jetpack.jp.ngrok.io/id-suscipit-aspernatur-nihil/?jb-generate-critical-css=5c2be69a79",
        "type": "CrossDomainError",
        "meta": {
          "url": "http://jetpack.jp.ngrok.io/id-suscipit-aspernatur-nihil/?jb-generate-critical-css=5c2be69a79"
        }
      },
      "http://jetpack.jp.ngrok.io/ut-aut-officia-necessitatibus-voluptatem-nostrum/": {
        "message": "Failed to fetch cross-domain content at http://jetpack.jp.ngrok.io/ut-aut-officia-necessitatibus-voluptatem-nostrum/?jb-generate-critical-css=5c2be69a79",
        "type": "CrossDomainError",
        "meta": {
          "url": "http://jetpack.jp.ngrok.io/ut-aut-officia-necessitatibus-voluptatem-nostrum/?jb-generate-critical-css=5c2be69a79"
        }
      }
    }
  },
  "provider_key_labels": {
    "singular_post": "Single post view"
  },
  "created": 1631792606.213865,
  "viewports": [
    {
      "type": "phone",
      "width": 414,
      "height": 896
    },
    {
      "type": "tablet",
      "width": 1200,
      "height": 800
    },
    {
      "type": "desktop",
      "width": 1920,
      "height": 1080
    }
  ],
  "generation_nonce": "5c2be69a79",
  "callback_passthrough": {
    "_nonce": "9bf70c6de2"
  }
}
```

If there is a show-stopper error (in this case noted by a `status` being `success` but `success_count` being `0`. This is actually counter-intuitive and we might want to fix it. Also, a `status` being `error` would be considered as a show-stopper). See [here](https://github.com/Automattic/jetpack/blob/add/track-critical-css/projects/plugins/boost/app/assets/src/js/pages/settings/elements/GenerateCSS.svelte#L15)


```
{
  "status": "success",
  "success_count": 0,
  "core_providers": [
    "core_front_page",
    "core_posts_page",
    "singular_page",
    "singular_post",
    "singular_product"
  ],
  "core_providers_status": "error",
  "providers_errors": {
    "core_posts_page": {
      "http://jetpack.jp.ngrok.io/": {
        "message": "Failed to verify page at http://jetpack.jp.ngrok.io/?jb-generate-critical-css=5c2be69a79",
        "type": "UrlVerifyError",
        "meta": {
          "url": "http://jetpack.jp.ngrok.io/?jb-generate-critical-css=5c2be69a79"
        }
      }
    },
    "singular_post": {
      "http://jetpack.jp.ngrok.io/id-suscipit-aspernatur-nihil/": {
        "message": "Failed to fetch cross-domain content at http://jetpack.jp.ngrok.io/id-suscipit-aspernatur-nihil/?jb-generate-critical-css=5c2be69a79",
        "type": "CrossDomainError",
        "meta": {
          "url": "http://jetpack.jp.ngrok.io/id-suscipit-aspernatur-nihil/?jb-generate-critical-css=5c2be69a79"
        }
      },
      "http://jetpack.jp.ngrok.io/ut-aut-officia-necessitatibus-voluptatem-nostrum/": {
        "message": "Failed to fetch cross-domain content at http://jetpack.jp.ngrok.io/ut-aut-officia-necessitatibus-voluptatem-nostrum/?jb-generate-critical-css=5c2be69a79",
        "type": "CrossDomainError",
        "meta": {
          "url": "http://jetpack.jp.ngrok.io/ut-aut-officia-necessitatibus-voluptatem-nostrum/?jb-generate-critical-css=5c2be69a79"
        }
      },
      "http://jetpack.jp.ngrok.io/quaerat-ut-nemo-maiores-facere-ut/": {
        "message": "Failed to verify page at http://jetpack.jp.ngrok.io/quaerat-ut-nemo-maiores-facere-ut/?jb-generate-critical-css=5c2be69a79",
        "type": "UrlVerifyError",
        "meta": {
          "url": "http://jetpack.jp.ngrok.io/quaerat-ut-nemo-maiores-facere-ut/?jb-generate-critical-css=5c2be69a79"
        }
      }
    },
    "taxonomy_category": {
      "http://jetpack.jp.ngrok.io/category/cupcake/": {
        "message": "Failed to verify page at http://jetpack.jp.ngrok.io/category/cupcake/?jb-generate-critical-css=5c2be69a79",
        "type": "UrlVerifyError",
        "meta": {
          "url": "http://jetpack.jp.ngrok.io/category/cupcake/?jb-generate-critical-css=5c2be69a79"
        }
      },
      "http://jetpack.jp.ngrok.io/category/dolorum/": {
        "message": "Failed to verify page at http://jetpack.jp.ngrok.io/category/dolorum/?jb-generate-critical-css=5c2be69a79",
        "type": "UrlVerifyError",
        "meta": {
          "url": "http://jetpack.jp.ngrok.io/category/dolorum/?jb-generate-critical-css=5c2be69a79"
        }
      },
      "http://jetpack.jp.ngrok.io/category/suscipit/": {
        "message": "Failed to verify page at http://jetpack.jp.ngrok.io/category/suscipit/?jb-generate-critical-css=5c2be69a79",
        "type": "UrlVerifyError",
        "meta": {
          "url": "http://jetpack.jp.ngrok.io/category/suscipit/?jb-generate-critical-css=5c2be69a79"
        }
      }
    }
  },
  "provider_key_labels": {
    "core_posts_page": "Posts page",
    "singular_post": "Single post view",
    "taxonomy_category": "Category view"
  },
  "created": 1631793163.939553,
  "viewports": [
    {
      "type": "phone",
      "width": 414,
      "height": 896
    },
    {
      "type": "tablet",
      "width": 1200,
      "height": 800
    },
    {
      "type": "desktop",
      "width": 1920,
      "height": 1080
    }
  ],
  "generation_nonce": "5c2be69a79",
  "callback_passthrough": {
    "_nonce": "9bf70c6de2"
  }
}
```

As you can see the final Critical CSS status already contains quite a lot of information that would be useful for the stats all at once:
- status
- success_count
- providers_errors

Now it does miss the Critical CSS size you are capturing as well as the generation time for each provider (very useful) and for the whole process.

I am wondering if we could actually store this information (Critical CSS size and providers and whole process generation time) inside the Critical CSS status and send that (or a part of that object as we probably don't need properties like `provider_key_labels`, `generation_nonce`, `core_providers` sent to Tracks.

I am thinking ideally we could have only one final object sent to tracks which could look like:

```
{
  "status": "success",
  "success_count": 2,
  "core_providers_status": "error",
  "providers": {
    "core_posts_page": {
      "urls": [
        "http://jetpack.jp.ngrok.io"
      ],
      "generation_time": 1234
    },
    "singular_post": {
      "urls": [
        "http://jetpack.jp.ngrok.io/id-suscipit-aspernatur-nihil/",
        "http://jetpack.jp.ngrok.io/ut-aut-officia-necessitatibus-voluptatem-nostrum/",
        "http://jetpack.jp.ngrok.io/quaerat-ut-nemo-maiores-facere-ut/"
      ],
      "generation_time": 10235
    },
    "taxonomy_category": {
      "urls": [
        "http://jetpack.jp.ngrok.io/category/cupcake/",
        "http://jetpack.jp.ngrok.io/category/dolorum/",
        "http://jetpack.jp.ngrok.io/category/suscipit/"
      ],
      "generation_time": 20122
    }
  },
  "providers_errors": {
    "singular_post": {
      "http://jetpack.jp.ngrok.io/id-suscipit-aspernatur-nihil/": {
        "message": "Failed to fetch cross-domain content at http://jetpack.jp.ngrok.io/id-suscipit-aspernatur-nihil/?jb-generate-critical-css=5c2be69a79",
        "type": "CrossDomainError",
        "meta": {
          "url": "http://jetpack.jp.ngrok.io/id-suscipit-aspernatur-nihil/?jb-generate-critical-css=5c2be69a79"
        }
      },
      "http://jetpack.jp.ngrok.io/ut-aut-officia-necessitatibus-voluptatem-nostrum/": {
        "message": "Failed to fetch cross-domain content at http://jetpack.jp.ngrok.io/ut-aut-officia-necessitatibus-voluptatem-nostrum/?jb-generate-critical-css=5c2be69a79",
        "type": "CrossDomainError",
        "meta": {
          "url": "http://jetpack.jp.ngrok.io/ut-aut-officia-necessitatibus-voluptatem-nostrum/?jb-generate-critical-css=5c2be69a79"
        }
      }
    }
  },
  "created": 1631792606.213865,
  "generation_time": 23865,
  "viewports": [
    {
      "type": "phone",
      "width": 414,
      "height": 896
    },
    {
      "type": "tablet",
      "width": 1200,
      "height": 800
    },
    {
      "type": "desktop",
      "width": 1920,
      "height": 1080
    }
  ]
}
```

Maybe this could be done partly where you are already doing it at https://github.com/Automattic/jetpack/blob/add/track-critical-css/projects/plugins/boost/app/assets/src/js/utils/generate-critical-css.ts#L251 but also if there is a show-stopper error at https://github.com/Automattic/jetpack/blob/add/track-critical-css/projects/plugins/boost/app/assets/src/js/utils/generate-critical-css.ts#L245

What do you think? Maybe this is out of the scope of this original ticket? cc @thingalon

davidlonjon commented on projects/plugins/boost/app/assets/src/js/utils/generate-critical-css.ts:
```diff
@@ -210,6 +248,16 @@ async function generateForKeys(
+	const { progress, percent_complete } = getCurrentStatus();
+	const trackerAttributes: TrackerAttributes = {
+		status: 'success',
+		progress,
```
I think at this stage, progress and percent_complete are not yet fully updated as shown in the screenshot below:

<img width="1535" alt="Screen Shot 2021-09-16 at 8 31 25 PM" src="https://user-images.githubusercontent.com/927932/133612886-011a5409-277e-405f-bc03-c975546de542.png">
Maybe also they are irrelevant in the case of success. We assume that the process as completed?

davidlonjon replied:
We also might want to send a final Track before the `return` at line 246 in case of a show-stopper error?

davidlonjon replied:
Finally, I am wondering it could be useful to send the provider_keys in here so have more details on the providers

thingalon replied:
A list of provider keys does sound useful. :)

ebinnion replied:
What are provider_keys?

thingalon replied:
> What are provider_keys?

When generating Critical CSS, we slice the user's site up into areas which are likely to use the same templates / CSS files. Each slice is identified by a short string called a "provider_key" (for internal / historic reasons).

davidlonjon commented on projects/plugins/boost/app/assets/src/js/utils/generate-critical-css.ts:
```diff
@@ -185,7 +201,29 @@ async function generateForKeys(
 			}
+
+			// Track individual Critical CSS generation success.
+			const trackerAttributes: TrackerAttributes = {
```
I am wondering if it would be useful to also send the URLs for a provider?

thingalon replied:
That data would be useful for debugging, but I'm not sure of the privacy implications of keeping that data. @ebinnion how do we feel about including URLs in tracking data?

ebinnion replied:
In this case, since the data is useful for debugging purposes, which will also assist us with helping our users, this seems reasonable to me.

davidlonjon commented on projects/plugins/boost/app/assets/src/js/utils/generate-critical-css.ts:
```diff
@@ -185,7 +201,29 @@ async function generateForKeys(
+			jpTracksAJAX.record_ajax_event( 'critical_css_success', 'click', trackerAttributes );
 		} catch ( err ) {
+			// Track individual Critical CSS generation error.
+			const trackerAttributes: TrackerAttributes = {
```
I am wondering if it would be useful to also send the URLs for a provider?

davidlonjon commented on projects/plugins/boost/app/assets/src/js/utils/generate-critical-css.ts:
```diff
@@ -185,7 +201,29 @@ async function generateForKeys(
+				error_message: err.message,
+			};
+			if ( err.isSuccessTargetError ) {
+				trackerAttributes.error_type = 'successTargetError';
```
I am wondering if we might want to send the `urlErrors` in the case of `isSuccessTargetError`?

thingalon replied:
I definitely want to include the type of URLError objects - but I'm not sure if we want it in [one big list, or as separate events](https://github.com/Automattic/jetpack/pull/21052#issuecomment-921381237).

davidlonjon reviewed (commented):
Hi @karthikax, I have added some comments inline with your code but they might or might not be relevant based on the larger comment I left you.

I hope it makes sense.

thingalon commented:
> If we don't take into consideration the size of the CSS generated and the time it took to generate it just yet (please bear with me on these two) I was wondering if it would be better to send to tracks only the final Critical CSS status so we have only one object for each site for each generation. For example, at the moment the final Critical CSS status can look like the following:

🤔  This is an interesting idea in spirit, but I don't want to tie Critical CSS status directly to the events we track for a few reasons. I want us to be able to update the Critical CSS status objects in the future to better suit / support the generation process without worrying about how it will affect tracking data collected. There are also details in it we don't need, and other details which aren't included that we want.

However, I like your suggestion to focus on a single tracking event at the end of a successful run so that each success is a single event in our system. (And to reduce the amount of data being thrown around).

@karthikax, when we discussed this idea originally I did suggest that we should track the size of each CSS block created, and that we could do that by sending an event-per-block but that does seem wasteful / over-the-top in retrospect. On further reflection (and after considering @davidlonjon's feedback), I'm going to suggest that we only send one object at the end of a standard successful Critical CSS generation run, which contains:

- The time it took to complete,
- The number of CSS blocks generated,
- The average size of CSS blocks generated,
- The size of the largest CSS block.

As for error tracking, this is a little bit more complex. I want to be able to know how often each type of URL error occurs (e.g.: "how often are users running into CrossDomainErrors") so that we can focus our dev efforts on making the workflow easier for users facing those problems -- and also "how often is the whole process a failure?" so that we can track our success rate.

My gut instinct is to track that by sending:

- 1 event every time a URL error occurs (i.e.: every time a single block / provider key of CSS fails to generate) with the type of error encountered, and
- 1 event at the end of every CSS generation attempt which totally fails (i.e.: with a showstopper error).

If we just bundle URL errors with a single "success" or "failure" event, I'm not sure how easy it would be for us to track stats on each type of error across all generation attempts. @ebinnion - would you mind weighing in on this?

If we can get away with one event per Critical CSS generation run - either a "success" or a "failure" event, that'd be great. Provided we can then break the data down into an overview of the most common _types_ of URL error which occurs during each path.

thingalon commented on projects/plugins/boost/app/assets/src/js/utils/generate-critical-css.ts:
```diff
@@ -154,10 +166,14 @@ async function generateForKeys(
 	const majorSteps = Object.keys( providerKeys ).length + 1;
 	let majorStep = 0;
 
+	// eslint-disable-next-line
```
Please always specify which linting rule is being disabled. e.g.:

```suggestion
	// eslint-disable-next-line @wordpress/no-unused-vars-before-return
```

It makes it clearer why the exception exists when others are reading the code later.

🤔  TBH I question the value of this linting rule, but that can/should be discussed outside the context of this PR.

thingalon commented on projects/plugins/boost/app/assets/src/js/utils/generate-critical-css.ts:
```diff
@@ -185,7 +201,29 @@ async function generateForKeys(
+				size: css.length,
+				generation_time: Date.now() - providerStartTime,
+			};
+			jpTracksAJAX.record_ajax_event( 'critical_css_success', 'click', trackerAttributes );
```
Is this a `click` event? I mean, I guess it happens well after a user clicks - but is there a more descriptive type we should use?

karthikax replied:
Current implementation of jetpack-tracking tracks the event-props only if it is a `click` event type. I think this needs to be updated to add more meaningful event types though!
Reference: https://github.com/Automattic/jetpack/blob/master/projects/packages/tracking/src/class-tracking.php#L77

thingalon reviewed (changes requested):
Off to a good start, @karthikax. I think I'd like a touch of advice from someone more knowledgeable than me in the appropriate shape to use for some of these Tracks events, so some of my feedback is pending some further input from @ebinnion

davidlonjon commented:
Hi @karthikax, I think @thingalon suggestions and comments are totally spot on you can carry on along his line :)

karthikax committed:
Add descriptive eslint exception rule.

karthikax committed:
Update Critical CSS generation stats design

Updated design contains better event props to track the statistics as discussed here: https://jetpackdatap2.wordpress.com/2021/09/17/tracking-critical-css-generation-statistics/

thingalon commented on projects/plugins/boost/app/assets/src/js/utils/generate-critical-css.ts:
```diff
@@ -205,11 +228,31 @@ async function generateForKeys(
+					time: Date.now() - startTime,
+					provider_key: providerKey,
+					error_message: err.message,
+					error_type: err.type,
```
Show-stopper errors won't be a `SuccessTargetError`, and may or may not be a `UrlError` subclass, so `type` may not be populated.

In this context, can I suggest something like:
```suggestion
					error_type: err.type || err.constructor && err.constructor.name || 'unknown'
```

(Only applies to show-stoppers; URL errors are guaranteed to have a `type` to use)

thingalon commented on projects/plugins/boost/app/assets/src/js/utils/generate-critical-css.ts:
```diff
@@ -205,11 +228,31 @@ async function generateForKeys(
+		time: Date.now() - startTime,
+		block_count: stepsPassed,
+		error_count: stepsFailed,
+		average_size: totalSize / stepsPassed,
```
```suggestion
		average_size: totalSize / Math.max( 1, stepsPassed ),
```

... just to be cautious and avoid divide-by-zero errors.

thingalon reviewed (changes requested):
This is really coming together 👏

karthikax committed:
Update eventProps type definitions

karthikax committed:
Merge branch 'master' into add/track-critical-css

karthikax commented:
Thanks, @davidlonjon, and @thingalon for the review.
Updated this PR with the changes requested. Tested it in the dashboard.
Also, updated the structure according to the discussion: panfyZ-en-p2

Updated events example:
![success](https://user-images.githubusercontent.com/8727183/134132166-c4ab9a9d-6dd0-4fac-8624-f55600f038d8.png)
![url_error](https://user-images.githubusercontent.com/8727183/134132189-7107f17c-1794-4792-81f8-973f337b56f6.png)

thingalon reviewed (approved):
Nice work. There is a function I'm not sure we need at the moment, but I don't think it's a blocker. 🚢

thingalon commented:
One further note: events are coming through as `jetpack_critical_css_*` - I would prefer `jetpack_boost_critical_css_*` instead so that our events are all namespaced :)

thingalon reviewed (changes requested):
I just discovered a bug in this.

There is a special case in the UI - if the process to generate Critical CSS succeeds, but _every provider key_ fails, it considers the process a failure. It's kind of a "showstopper error by way of having all the minor errors".

You can reproduce it with a bit of code which throws an HTTP error whenever the `jb-generate-critical-css` GET parameter is used. It will cause all the generation steps to fail -- but it will end up logging a success in the message sent to tracks.

karthikax commented on projects/plugins/boost/app/assets/src/js/stores/critical-css-status.ts:
```diff
@@ -63,6 +65,13 @@ const initialState = Jetpack_Boost.criticalCssStatus || {
+// Keep a subscribed copy for quick reading.
+let currentStatus: CriticalCssStatus;
+subscribe( value => ( currentStatus = value ) );
+export function getCurrentStatus(): CriticalCssStatus {
```
No, It is no more relevant. Thanks

karthikax committed:
Remove redundant function getCurrentStatus

karthikax committed:
Add overall Critical CSS error tracking

In a few edge cases, Critical CSS generation results in failure for all the provider keys. Added event tracking for the same.

karthikax committed:
Merge branch 'master' into add/track-critical-css

karthikax committed:
Add Tracks ajax filter

kbrown9 committed:
Tracking: set up the ajax hook in the Tracking class

* Set up the ajax hook in the Tracking class. Use an action to ensure that the hook is set up only once.
* Since the hook is set up once, the event type prefix will be the product name of the object that set up the hook. This
  isn't the expected behavior. Fix that by adding a new parameter to the record_user_event() method, 'use_product_prefix'.
  This parameter is used to indicate whether the product name should be used as the prefix. If false, 'jetpack' is used.
* In the 'ajax_tracks' method, pass 'false' as the value of 'use_product_prefix' in the call to record_user_event(). For
  events that are tracked via Ajax, the prefix will always be 'jetpack_', as indicated in the README document.

karthikax committed:
Merge branch 'master' into test/tracks-ajax

karthikax committed:
Revert "Add Tracks ajax filter"

This reverts commit adc046a195f11978341daf1c3edd2e60da4e6950.

karthikax committed:
Add boost prefix for tracks event

karthikax committed:
Merge branch 'master' into test/tracks-ajax

karthikax committed:
Merge branch 'test/tracks-ajax' into add/track-critical-css

karthikax committed:
Remove Jetpack specific Tracks ajax hook

Since ajax hook is defined in the Tracking class now, defining wp_ajax_jetpack_tracks in a consumer, creates duplication and hence removed.

karthikax committed:
Add jetpack changelog

karthikax commented:
Based on the discussion (p9dueE-3w6-p2), to avoid the event name conflict between Jetpack and Jetpack Boost, the universal ajax callback is now moved to the Tracking package.

Thanks a lot, @kbrown9 @thingalon @leogermani @jeherve for the assistance.

Kim, I have cherry-picked your commit to this PR. Please let me know if this needs to go to a separate one!

I have tested and verified that the implemented solution works.
- Tested with different combinations of Jetpack and Jetpack Boost activated and deactivated states.
- Tested from both the local environment and with a proxy tunnel.
- Tested the events from both the plugins.

A few screenshots for reference:
<img width="1349" alt="Screenshot" src="https://user-images.githubusercontent.com/8727183/134920720-c3931938-1769-4a59-8dab-534a7cc24816.png">

karthikax committed:
Revert "Tracking ajax changes"

Revert following commits: a4a005a, 096d124, b008d31

karthikax committed:
Merge branch 'master' into add/track-critical-css

karthikax commented:
Reverted the Jetpack plugin-specific Tracking changes from this PR and moved them to a [separate PR](https://github.com/Automattic/jetpack/pull/21211), since both these PRs address separate concerns.

Note: This PR to be merged only after we merge #21211

karthikax committed:
Remove tracks ajax callback within Boost analytics

thingalon reviewed (approved):
👍  This is looking great. I know it can't go out before https://github.com/Automattic/jetpack/pull/21211 but as soon as 21211 is in, this can go in too.

karthikax committed:
Merge branch 'master' into add/track-critical-css
